Documented ingeter migration to spread-minimizing tokens (#7174)
grafana/mimir: docs/sources/mimir/configure/configure-spread-minimizing-tokens/_index.md

---
aliases:
  - ../operators-guide/configure/configure-spread-minimizing-tokens/
description: Learn how to migrate ingesters to spread-minimizing tokens.

> weight: 115

Reviewer: I added a page weight to control where this page appears in the left table of contents. With this change, Spread-minimizing tokens appears after Zone-aware replication. 
![image](https://github.com/grafana/mimir/assets/104772500/4c4c624f-f886-4ad6-a014-d9e268f57124)

menuTitle: Spread-minimizing tokens
title: Migrate ingesters to spread-minimizing tokens
---

# Migrate ingesters to spread-minimizing tokens

> Using this guide, you can configure Mimir's ingesters to use the _spread-minimizing token generation strategy_ .

Reviewer: You have a good open. I tried to remove the passive voice. What do you think? Also, maybe include a sentence about why someone would want to use this and when. 

The suggestion below just removes an extra space before the period. 
```suggestion
Using this guide, you can configure Mimir's ingesters to use the _spread-minimizing token generation strategy_.
```

## Before you begin

The ingester time series replication should be [configured with enabled zone-awareness](https://grafana.com/docs/mimir/latest/configure/configure-zone-aware-replication/#configuring-ingester-time-series-replication).

> {{% admonition type="note" %}}Use spread-mimizing tokens are recommended if [shuffle sharding](https://grafana.com/docs/mimir/latest/configure/configure-shuffle-sharding/#ingesters-shuffle-sharding) is disabled in your ingesters, or, if shuffle-sharding is enabled, but most of the tenants of your system use all available ingesters.

Reviewer: I replaced the notes with the admonition shortcode in your document.  

Notes, tips, warnings, and caution admonitions appear with special formatted boxes when rendered. The [admonition shortcode](https://grafana.com/docs/writers-toolkit/write/shortcodes/) tells Hugo to set this paragraph apart with different formatting. 

Here's what the above admonition shortcode looks like in the doc preview: 
![image](https://github.com/grafana/mimir/assets/104772500/c1c78a80-e107-4798-9a54-b95e2f504ea5)

Reviewer: and I found another mistake in this sentence. Corrected below. 
```suggestion
{{% admonition type="note" %}}Spread-mimizing tokens are recommended if [shuffle sharding](https://grafana.com/docs/mimir/latest/configure/configure-shuffle-sharding/#ingesters-shuffle-sharding) is disabled in your ingesters, or, if shuffle-sharding is enabled, but most of the tenants of your system use all available ingesters.
```

{{% /admonition %}}

For simplicity, let’s assume that there are three configured availability zones named `zone-a`, `zone-b`, and `zone-c`. Migration to the _spread-minimizing token generation strategy_ is a complex process performed zone by zone to prevent any data loss.

> ## Step 1: Disable write requests to ingesters from zone-a

Reviewer: ```suggestion
## Step 1: Disable write requests to ingesters from `zone-a`
```

> To disable write requests, configure the following flag on the distributor and the ruler:

Reviewer: I removed most of the uses of "in order to" and instead used just "to".

```
-ingester.ring.excluded-zones=zone-a
```

Before proceeding to the next step, use the following query to ensure that there are no write requests to the ingesters from `zone-a`:

```
sum by(route) (
  rate(
    cortex_request_duration_seconds_count{
      namespace="<your-namespace>",
      container="ingester",
      pod=~"ingester-zone-a-.*",
      route="/cortex.Ingester/Push"}[5m]
  )
)
```

You should see something like this:

![No More Write Requests](no-more-write-requests.png)

> ## Step 2: Shut down ingesters from zone-a

Reviewer: I missed this. Zone-a should have formatting, same as you've used in the text. 
```suggestion
## Step 2: Shut down ingesters from `zone-a`
```

Next, ensure that all the in-memory series of all ingesters from `zone-a` have been flushed to long-term storage, as well as that the ingesters from `zone-a` have been forgotten from the ring.
To do this, invoke the [/ingester/shutdown](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/references/http-api/index.md#shutdown) endpoint on all the ingesters from `zone-a`.

Before proceeding to the next step, ensure that all the calls completed successfully completed.

> ## Step 3: Enable spread-minimizing tokens for ingesters in zone-a

Reviewer: ```suggestion
## Step 3: Enable spread-minimizing tokens for ingesters in `zone-a`
```

Configure the following flags on the ingesters from `zone-a`:

```
-ingester.ring.tokens-file-path=
-ingester.ring.token-generation-strategy=spread-minimizing
-ingester.ring.spread-minimizing-zones=zone-a,zone-b,zone-c
```

{{% admonition type="note" %}}
The example uses `zone-a,zone-b,zone-c` to denote a comma-separated list of configured availability zones.
{{% /admonition %}}

Before proceeding to the next step, ensure that all the ingester pods related to `zone-a` are up and running with the new configuration.

### Optional step: In-order registration of ingesters

> Mimir can force the ring to perform an in-order registration of ingesters, for example, to allow an ingester to register its tokens within the ring only after all previous ingesters (with ID lower than its own ID) have already been registered.

Reviewer: Lines 74-76 had some run-on sentences with some complicated ideas. I tried to split them into several sentences. Does this look okay?

Enabling this feature minimizes a probability that a write request that should be handled by an ingester actually arrives to the ring before the ingester is registered within the ring. In this case, the request gets handled by another ingester.

> This situation that would introduce some deviation from an optimal load distribution.

Reviewer: ```suggestion
This situation could introduce some deviation from an optimal load distribution.
```
